File: workspace/src/py_image_processor/py_image_processor/image_processor_node.py
import time
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
import cv2
from cv_bridge import CvBridge
import numpy as np

class ImageProcessorNode(Node):
    def __init__(self):
        super().__init__('image_processor_node')
        self.logger = rclpy.logging.get_logger(self.get_name())

        self.image_sub = self.create_subscription(Image, "image/original", self.image_callback, 10)
        self.image_sub # avoid unused variable warning

        self.image_pub = self.create_publisher(Image, "image/grayscale", 1)

        self.clock_ = rclpy.clock.Clock()
        self.bridge = CvBridge()

    def image_callback(self, msg):
        cur_time = self.clock_.now()
        secs = cur_time.nanoseconds//1000000000 - msg.header.stamp.sec
        nsecs = cur_time.nanoseconds%1000000000 - msg.header.stamp.nanosec
        self.get_logger().info(f"Time diff: {secs}.{nsecs}s")

        cv_image = self.bridge.imgmsg_to_cv2(msg)
        gray = cv2.cvtColor(cv_image, cv2.COLOR_BGR2GRAY)       
        gray_msg = self.bridge.cv2_to_imgmsg(gray)

        gray_msg.header.stamp = self.clock_.now().to_msg()
        self.image_pub.publish(gray_msg)
        

def main(args=None):
    rclpy.init(args=args)
    image_processor_node = ImageProcessorNode()

    rclpy.spin(image_processor_node)

    image_processor_node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
        

File: workspace/src/archive/cpp_image_processor/src/image_processor_node.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "sensor_msgs/msg/image.hpp"
#include <opencv2/opencv.hpp>
#include <cv_bridge/cv_bridge.h>

class ImageProcessorNode : public rclcpp::Node
{
	public:
		ImageProcessorNode() : Node("image_processor_node"), count_(0), clock_()
		{
			image_sub = this->create_subscription<sensor_msgs::msg::Image>("image/original", 10, std::bind(&ImageProcessorNode::image_callback, this, std::placeholders::_1));
            
            image_pub = this->create_publisher<sensor_msgs::msg::Image>("image/grayscale", 10);
		}
	private:
		void image_callback (const sensor_msgs::msg::Image::SharedPtr msg) 
		{

			rclcpp::Time cur_time = clock_.now();
			const auto secs = cur_time.nanoseconds()/1000000000UL - msg->header.stamp.sec;
			const auto nsecs = cur_time.nanoseconds()%1000000000UL - msg->header.stamp.nanosec;

			std::stringstream ss;
			ss << "Time diff: " << secs << "." << nsecs << "s";
			RCLCPP_INFO(this->get_logger(), ss.str().c_str());

            // convert Image msg to CvImagePtr
            cv_bridge::CvImagePtr colorPtr;
            colorPtr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
            
            // get grayscale image
            cv::Mat colorMat = colorPtr->image;
            cv::Mat grayMat;
            cv::cvtColor(colorMat, grayMat, CV_BGR2GRAY);

            // convert cv::Mat to CvImage
            std_msgs::msg::Header header; // empty header
            header.stamp = clock_.now();
            cv_bridge::CvImage grayPtr;
            grayPtr = cv_bridge::CvImage(header, sensor_msgs::image_encodings::MONO8, grayMat);

            // publish
            image_pub->publish(*grayPtr.toImageMsg());
		}

		rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr image_sub;
        rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr image_pub;
		size_t count_;
		rclcpp::Clock clock_;

};

int main(int argc, char* argv[])
{
	rclcpp::init(argc, argv);
	rclcpp::spin(std::make_shared<ImageProcessorNode>());
	rclcpp::shutdown();
	return 0;
}
